lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -837,95 +947,119 @@ func (c *ContainerProxy) Resize(vc *viccontainer.VicContainer, height, width int
 
 // AttachStreams takes the the hijacked connections from the calling client and attaches
 // them to the 3 streams from the portlayer's rest server.
-// clStdin, clStdout, clStderr are the hijacked connection
-func (c *ContainerProxy) AttachStreams(ctx context.Context, vc *viccontainer.VicContainer, clStdin io.ReadCloser, clStdout, clStderr io.Writer, ca *backend.ContainerAttachConfig) error {
+// stdin, stdout, stderr are the hijacked connection
+func (c *ContainerProxy) AttachStreams(ctx context.Context, ac *AttachConfig, stdin io.ReadCloser, stdout, stderr io.Writer) error {
 	// Cancel will close the child connections.
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	var wg sync.WaitGroup
+	var wg, outWg sync.WaitGroup
 	errors := make(chan error, 3)
 
 	// For stdin, we only have a timeout for connection.  There can be a long duration before
 	// the first entry so there is no timeout for response.
-	plClient, transport := c.createNewAttachClientWithTimeouts(attachConnectTimeout, 0, attachAttemptTimeout)
+	client, transport := c.createNewAttachClientWithTimeouts(attachConnectTimeout, 0, attachAttemptTimeout)
 	defer transport.Close()
 
 	var keys []byte
 	var err error
-	if ca.DetachKeys != "" {
-		keys, err = term.ToBytes(ca.DetachKeys)
+	if ac.DetachKeys != "" {
+		keys, err = term.ToBytes(ac.DetachKeys)
 		if err != nil {
-			return fmt.Errorf("Invalid escape keys (%s) provided", ca.DetachKeys)
+			return fmt.Errorf("Invalid escape keys (%s) provided", ac.DetachKeys)
 		}
 	}
 
-	if ca.UseStdin {
+	// ensure that the API client connection is shutdown on exit
+	// TODO: see if this is needed, or if callers of postContainersAttach will clean up
+	// on clean exit path
+	//defer func() {
```

Comment:
dropping it, we need streams open after returning this function so that we can report errors back to the client